esp/main/src/TemperatureSensor.cpp

```diff
@@ -0,0 +1,127 @@
+        if (searchState.rom_code == knownDevices[0] ||
+            searchState.rom_code == knownDevices[1]) {
+            //            device_rom_codes[NoDevices] = searchState.rom_code;
+            NoDevices++;
+        }
+        owb_search_next(oneWireInterface, &searchState, &found);
+        totalDevicesNo++;
+    }
+    if (NoDevices != totalDevicesNo) {
+        std::cout << "Unwanted devices detected!!! Total NoDevices: "
+                  << totalDevicesNo << std::endl;
+    }
+    std::cout << "Found " << NoDevices << "device"
+              << (NoDevices == 1 ? "" : "s") << std::endl;
+    if (NoDevices != 2)
+        esp_restart();
+}
+
+void TemperatureSensor::InitializeDevices() {
+    for (int i = 0; i < NoDevices; i++) {
+        DS18B20_Info* ds18b20_info = ds18b20_malloc(); // heap allocation
+        devices[i] = ds18b20_info;
+        ds18b20_init(ds18b20_info,
+                     oneWireInterface,
+                     knownDevices[i]); // associate with bus and device
+        ds18b20_use_crc(ds18b20_info, true); // enable CRC check on all reads
+        ds18b20_set_resolution(ds18b20_info, DS18B20_RESOLUTION);
+    }
+}
+
+// Won't be used in final version
+void TemperatureSensor::Run() {
+    TickType_t lastWakeTime;
+    if (NoDevices > 0) {
+        while (true) {
+            lastWakeTime = xTaskGetTickCount();
+            ds18b20_convert_all(oneWireInterface);
+            ds18b20_wait_for_conversion(devices[0]);
+            auto readings = ReadTemperature();
+            if (IsErrorInReading()) {
+                break;
+            }
+            DisplayTemperature(readings);
+            vTaskDelayUntil(&lastWakeTime, SAMPLE_PERIOD / portTICK_PERIOD_MS);
+        }
+    }
+    esp_restart();
+}
+
+std::string
+TemperatureSensor::CreateStringFromRom(OneWireBus_ROMCode device) const {
+    char rom_code_s[OWB_ROM_CODE_STRING_LENGTH];
+    owb_string_from_rom_code(device, rom_code_s, sizeof(rom_code_s));
+    return std::string(rom_code_s);
+}
+
+bool TemperatureSensor::IsErrorInReading() {
+    for (const DS18B20_ERROR err : errors) {
+        if (err != DS18B20_OK) {
+            return true;
+        }
+    }
+    return false;
+}
+
+std::array<float, MAX_DEVICES> TemperatureSensor::ReadTemperature() {
+    std::array<float, MAX_DEVICES> readings{};
+    for (int i = 0; i < NoDevices; ++i) {
+        errors[i] = ds18b20_read_temp(devices[i], &readings[i]);
+    }
+    return readings;
+}
+
+// Won't be used in final version
+template<std::size_t index>
+void TemperatureSensor::DisplayTemperature(
+    std::array<float, index> readings) const {
+    for (int i = 0; i < NoDevices; ++i) {
+        std::cout << "Temperature readings for "
+                  << TemperatureSensor::CreateStringFromRom(
+                         devices[i]->rom_code)
+                  << " : " << readings[i]++ << std::endl;
+    }
+}
+
+std::array<float, MAX_DEVICES> TemperatureSensor::PerformTemperatureReadOut() {
```

Comment:
auto

Reply:
It's cpp 11, although I'll use auto in case I ever switch to cpp 14.